test/integration/apiserver/cel/admission_test_util.go

```diff
@@ -0,0 +1,1084 @@
+	if len(request.SubResource) > 0 {
+		resource.Resource += "/" + request.SubResource
+	}
+
+	// See if we should record this
+	gvrToRecord := h.recordGVR
+	if converted {
+		// If this is a converted webhook, map to the GVR we expect the webhook to see
+		gvrToRecord = h.gvrToConvertedGVR[h.recordGVR]
+	}
+	if resource != gvrToRecord {
+		if debug {
+			h.t.Log(resource, "!=", gvrToRecord)
+		}
+		return
+	}
+
+	if request.Operation != h.recordOperation {
+		if debug {
+			h.t.Log(request.Operation, "!=", h.recordOperation)
+		}
+		return
+	}
+	if request.Namespace != h.recordNamespace {
+		if debug {
+			h.t.Log(request.Namespace, "!=", h.recordNamespace)
+		}
+		return
+	}
+
+	name := request.Name
+	if name != h.recordName {
+		if debug {
+			h.t.Log(name, "!=", h.recordName)
+		}
+		return
+	}
+
+	if debug {
+		h.t.Logf("recording: %#v = %s %#v %v", webhookOptions{version: version, phase: phase, converted: converted}, request.Operation, request.Resource, request.SubResource)
+	}
+	h.recorded[webhookOptions{version: version, phase: phase, converted: converted}] = request
+}
+
+func (h *holder) verify(t *testing.T) {
+	h.lock.Lock()
+	defer h.lock.Unlock()
+
+	for options, value := range h.recorded {
+		if err := h.verifyRequest(options, value); err != nil {
+			t.Errorf("version: %v, phase:%v, converted:%v error: %v", options.version, options.phase, options.converted, err)
+		}
+	}
+}
+
+func (h *holder) verifyRequest(webhookOptions webhookOptions, request *admissionRequest) error {
+	converted := webhookOptions.converted
+
+	// Check if current resource should be exempted from Admission processing
+	if h.admissionExemptResources[gvr(h.recordGVR.Group, h.recordGVR.Version, h.recordGVR.Resource)] {
+		if request == nil {
+			return nil
+		}
+		return fmt.Errorf("admission webhook was called, but not supposed to")
+	}
+
+	if request == nil {
+		return fmt.Errorf("no request received")
+	}
+
+	if h.expectObject {
+		if err := h.verifyObject(converted, request.Object.Object); err != nil {
+			return fmt.Errorf("object error: %v", err)
+		}
+	} else if request.Object.Object != nil {
+		return fmt.Errorf("unexpected object: %#v", request.Object.Object)
+	}
+
+	if h.expectOldObject {
+		if err := h.verifyObject(converted, request.OldObject.Object); err != nil {
+			return fmt.Errorf("old object error: %v", err)
+		}
+	} else if request.OldObject.Object != nil {
+		return fmt.Errorf("unexpected old object: %#v", request.OldObject.Object)
+	}
+
+	if h.expectOptions {
+		if err := h.verifyOptions(request.Options.Object); err != nil {
+			return fmt.Errorf("options error: %v", err)
+		}
+	} else if request.Options.Object != nil {
+		return fmt.Errorf("unexpected options: %#v", request.Options.Object)
+	}
+
+	//! Check for webhook tests only
+	// if !h.warningHandler.hasWarning(makeWarning(webhookOptions.version, webhookOptions.phase, webhookOptions.converted)) {
+	// 	return fmt.Errorf("no warning received from webhook")
+	// }
```

Comment:
It might not be clear to most users that the commented out code signals a difference between this code and the webhook test code. Clarify?